File: 2050-parallel-courses-iii/2050-parallel-courses-iii.java
public class Solution {
    public int minimumTime(int n, int[][] relations, int[] time) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        int[] in_degree = new int[n + 1];
        for (int[] relation : relations) {
            graph.computeIfAbsent(relation[0], k -> new ArrayList<>()).add(relation[1]);
            in_degree[relation[1]]++;
        }

        int[] dist = new int[n + 1];
        System.arraycopy(time, 0, dist, 1, n);
        Queue<Integer> q = new LinkedList<>();
        for (int i = 1; i <= n; i++) {
            if (in_degree[i] == 0) {
                q.add(i);
            }
        }

        while (!q.isEmpty()) {
            int course = q.poll();
            if (graph.containsKey(course)) {
                for (int next_course : graph.get(course)) {
                    dist[next_course] = Math.max(dist[next_course], dist[course] + time[next_course - 1]);
                    in_degree[next_course]--;
                    if (in_degree[next_course] == 0) {
                        q.add(next_course);
                    }
                }
            }
        }

        int maxVal = 0;
        for (int val : dist) {
            maxVal = Math.max(maxVal, val);
        }
        return maxVal;
    }
}

File: 2050-parallel-courses-iii/2050-parallel-courses-iii.js
/**
 * @param {number} n
 * @param {number[][]} relations
 * @param {number[]} time
 * @return {number}
 */
var minimumTime = function(n, relations, time) {
    const adjList = new Array(n + 1)
    const inDegree = new Array(n + 1).fill(0)
    const costs = new Array(n + 1).fill(0)
    for (let i = 0; i <= n; i++) {
        adjList[i] = [];
    }
    for (const [from, to] of relations) {
        inDegree[to] += 1
        adjList[from].push(to)
    }
    const q = []
    for (let i = 1; i <= n; i++) {
        if (inDegree[i] === 0) {
            costs[i] = time[i - 1]
            q.push(i) 
        }
    }

    while(q.length) {
        const u = q.shift();
        const cost = costs[u]
        const list = adjList[u] || []
        for (const v of list) {
            inDegree[v] -= 1
            costs[v] = Math.max(costs[v], cost)
            if (inDegree[v] === 0) {
                costs[v] += time[v - 1]
                q.push(v)
            }
        }
    }
    return Math.max(...costs)
};